gen/gen_enums.py

```diff
@@ -0,0 +1,131 @@
+#!/usr/bin/env python
+
+#
+# Quick & dirty component impl generator.
+# From the root of the EN tree, run:
+#    gen/gen_component_impl.py
+#
+# (Or, on windows: python gen/gen_component_impl.py)
+#
+
+import re
+import collections
+import sys
+import io
+import os
+import glob
+
+def fail(args):
+    print(args)
+    exit(1)
+
+def main():
+    files = [f for f in glob.glob('gen/enum/*') if not f.endswith('.py')]
+
+    pattern = re.compile(r'\s*(\w+)\s*(?:=\s*(\d+))?\s*')
+    tab = "    "
+
+    for fl in files:
+        print("Genning from %s to " % fl)
+        enum_name = fl.split(os.sep)[-1]
+        input_file = fl
+
+        output_dir = "src/enums/"
+
+        header_file = os.path.join(output_dir, enum_name + ".h")
+        source_file = os.path.join(output_dir, enum_name + ".cc")
+        print("  header: %s" % header_file)
+        print("  source: %s" % source_file)
+
+        next_value = 0
+
+        fields = collections.OrderedDict()
+
+        for line in open(input_file, 'r'):
+            line = line.strip()
+
+            if len(line) == 0:
+                continue
+
+            match = pattern.match(line)
+
+            if match is None:
+                fail("Syntax error: [" + line + "]")
+
+            field_name = match.group(1)
+
+            if field_name in fields:
+                fail("Double field name: " + field_name)
+
+            value_string = match.group(2)
+
+            if value_string is not None:
+                value = int(value_string)
+
+                if value < next_value:
+                    fail("Not a monotonous progression from " + next_value + " to " + value + " for enum field " + field_name)
+
+                next_value = value
+
+            fields[field_name] = next_value
+
+            next_value += 1
+
+        header_writer = open(header_file, 'w')
+        code_writer = open(source_file, 'w')
+
+        try:
```

Comment:
This is usually written `with open(...) as header_writer:` .. rather than explicit try/finally